Add pwndbg.storage and pwndbg.memoize.forever

zachriggle opened the pull request:
This is an alternative implementation to #268 that I contest is a simpler and more readable. 

Notable differences:

- It uses the built-in `shelve` module for persistence.
- It provides a volatile (tempfile) and non-volatile (~/.pwndbg) storage option
  - The path is always the same, regardless of OS.  Nice and simple, and the documentation is easier, too.
  - Exposes standard GDB settings (`pwndbg.config`) to control the locations if the user wants, rather than environment variables.
- Cache validity is determined by the modification time of the file that contains the function, rather than the function signature.  
  - This is more robust for development cycles where the logic may change, even if the signature doesn't.  
  - This still doesn't track changes in dependent routines.
  - There is actually an implied reliance on function signature, since we store the full set of arguments
- Adds `pwndbg.memoize.forever` and an alias to `pwndbg.storage.NonVolatile`

zachriggle committed:
Add pwndbg.storage and pwndbg.memoize.forever

zachriggle commented:
Actually, I should change this so that the top-level key is the full path of the function being cached

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,105 @@
+class Cached(object):
+    """Function decorator that persistently caches the results of the function it decorates."""
+    function = None
+    type = volatile
```
Right now `Cached` and `Volatile` does the same thing.

To prevent one from randomly choosing `Cached` over `Volatile`, the `Cached` should be an abstract class.

However, at least in Py2 it is not possible to have an ABC without a method or property:
```python
In [3]: class A(object):
   ...:     __metaclass__ = ABCMeta
   ...:     

In [4]: class B(A):
   ...:     pass
   ...: 

In [5]: A()
Out[5]: <__main__.A at 0x7fe4c3aa2610>

In [6]: B()
Out[6]: <__main__.B at 0x7fe4c3aac590>
```

So this can be fixed by moving `type` into an `abstractproperty`:
```python
In [19]: class A(object):
    ...:     __metaclass__ = ABCMeta
    ...:     @abstractproperty
    ...:     def type(self):
    ...:         pass
    ...:     

In [20]: class B(A):
    ...:     @property
    ...:     def type(self):
    ...:         return "This is B"
    ...:     

In [21]: A()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-21-3cd318a12eea> in <module>()
----> 1 A()

TypeError: Can't instantiate abstract class A with abstract methods type

In [22]: B().type
Out[22]: 'This is B'
```

zachriggle replied:
Nah, it's fine the way it is.  This greatly overcomplicates things to avoid the simple and harmless case where "Cached and Volatile are the same thing".

disconnect3d replied:
Hm.. and do we have any use case for more then `volatile` and `nonvolatile` type?

We could also go for having `volatile=True/False` as `Cached's __init__` argument.

zachriggle replied:
Done in ba99595

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,105 @@
+        functools.update_wrapper(self, function)
+
+    def __setup_shelf(self):
+        """Deferred initialization allows customization of the paths by the user"""
```
I'd add "through `type` property" (if we make it a property ;P).

zachriggle replied:
No, it's specifically so that the `pwndbg.config.Parameter` (which defines the filesystem path) can be changed by the user before we decide where to store the database

disconnect3d replied:
Suggested this because of previous suggestion of moving `type` into property :P.

zachriggle replied:
This is now set via the constructor, and `Volatile` and `NonVolatile` now just inherit from `Cached` and provide a default value

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,105 @@
+        # Cache per-architecture x arguments
+        args = inspect.getcallargs(self.function, *a, **kw)
+
+        # Dict is not hashable, convert into a tuple so it can be a dictk ey
```
Typo `dictk ey` -> `dict key`
So bad Python has no `FronzenDict` builtin...

zachriggle replied:
Quite unfortunate! :(

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,105 @@
+        self.shelf.setdefault('create_time', time.time())
+
+        # Get the modification time of the function's Python file
+        function_mtime = os.path.getmtime(inspect.getfile(self.function))
```
Just saying - `inspect.getfile` returns `<string>` for stuff wrapped in Python Interpreter:
```python
pwndbg> pi
In : from pwndbg.storage import Volatile, NonVolatile
In : @Volatile
...: def f():
...:     return 123
...: 
In : f()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 68, in __call__
    self.__setup_shelf()
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 58, in __setup_shelf
    function_mtime = os.path.getmtime(inspect.getfile(self.function))
  File "/usr/lib/python3.6/genericpath.py", line 55, in getmtime
    return os.stat(filename).st_mtime
FileNotFoundError: [Errno 2] No such file or directory: '<stdin>'
```

zachriggle replied:
That's fine.  Don't do that :)

zachriggle replied:
Fixed this so that if `os.path.mtime` fails for any reason, we set the modification time to 99999999999999 or so

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,105 @@
+        args = inspect.getcallargs(self.function, *a, **kw)
+
+        # Dict is not hashable, convert into a tuple so it can be a dictk ey
+        args = tuple(sorted(args.iteritems()))
```
`args.items()` (for a very big dict it is a performance hit on py2) or `six.iteritems(args)`

zachriggle replied:
Given that the size will be based on how many arguments there are for a function, we can safely assume the size doesn't matter.

zachriggle replied:
Fixed to use `items`

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,105 @@
+    func_name = function.__name__
+    return mod_name + '.' + func_name
+
+def path(function, directory=volatile):
```
Never called with the default.

zachriggle replied:
Meh :)

zachriggle replied:
This function is now gone

disconnect3d commented:
I agree this is more simple & readable.

The path `~/.pwndbg` is never created, and so using `NonVolatile` demo version booms:
```python
pwndbg> pi
In : demo()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 68, in __call__
    self.__setup_shelf()
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 52, in __setup_shelf
    protocol=pickle.HIGHEST_PROTOCOL)
  File "/usr/lib/python3.6/shelve.py", line 243, in open
    return DbfilenameShelf(filename, flag, protocol, writeback)
  File "/usr/lib/python3.6/shelve.py", line 227, in __init__
    Shelf.__init__(self, dbm.open(filename, flag), protocol, writeback)
  File "/usr/lib/python3.6/dbm/__init__.py", line 94, in open
    return mod.open(file, flag, mode)
_gdbm.error: [Errno 2] No such file or directory
```

zachriggle committed:
Fix some issues with storage and only use one instance each

zachriggle commented:
@disconnect3d I think that I fixed it.  I also simplified some of the logic -- there's now only two `Shelf` instances ever, which should fix some issues with multiple concurrent writes / state synchronization.

zachriggle committed:
Change the way volatile/nonvolatile is selected, and re-add architecture dependency

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,144 @@
+        database_path = os.path.join(str(db), 'storage')
+
+        # Ensure the path exists, then open the database
+        if not os.path.isdir(database_path):
```
The `database_path` must be a path to a file (might be non existing) otherwise `shelve.open` fails:
```
$ gdb
Defined pretty print function (pp) and libc (CDLL)
Loaded 108 commands. Type pwndbg [filter] for a list.
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/gdbinit.py", line 36, in <module>
    import pwndbg # isort:skip
  File "/home/dc/installed/pwndbg/pwndbg/__init__.py", line 69, in <module>
    import pwndbg.storage
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 53, in <module>
    update()
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 50, in update
    protocol=pickle.HIGHEST_PROTOCOL)
  File "/usr/lib/python3.6/shelve.py", line 243, in open
    return DbfilenameShelf(filename, flag, protocol, writeback)
  File "/usr/lib/python3.6/shelve.py", line 227, in __init__
    Shelf.__init__(self, dbm.open(filename, flag), protocol, writeback)
  File "/usr/lib/python3.6/dbm/__init__.py", line 94, in open
    return mod.open(file, flag, mode)
_gdbm.error: [Errno 21] Is a directory
```

I'd go with:
```python
for db in map(str, (volatile, nonvolatile)):
    ...
```

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,144 @@
+    global shelves
+
+    for db in (volatile, nonvolatile):
+        database_path = os.path.join(str(db), 'storage')
```
For volatile we get a path of `/tmp/storage`. It should be more obvious that the storage comes from pwndbg.

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,144 @@
+
+class Volatile(Cached):
+    def __init__(self, function):
+        super(NonVolatile, self).__init__(function, volatile)
```
Copy paste! `super(NonVolatile, self)` -> `super(Volatile, self)`

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,144 @@
+    def __init__(self, function, type):
+        self.function = function
+        self.function_name = function.__module__ + '.' + function.__name__
+        self.type = Type
```
`Type` is undefined. Also, `type` shadows the builtin, so if we do care, maybe `type_`

disconnect3d commented on pwndbg/storage.py:
```diff
@@ -0,0 +1,144 @@
+
+    @property
+    def initialized(self):
+        return self.function_name in self.shelf and 'create_time' in self.calls
```
`self.calls` is not defined

disconnect3d reviewed (changes requested):
Looks for me like WIP or you haven't pushed all commits :f

disconnect3d commented:
Uh.. there is a kinda big problem when you run more than one instance of pwndbg:
```
Loaded 108 commands. Type pwndbg [filter] for a list.
Traceback (most recent call last):
  File "/home/dc/installed/pwndbg/gdbinit.py", line 36, in <module>
    import pwndbg # isort:skip
  File "/home/dc/installed/pwndbg/pwndbg/__init__.py", line 69, in <module>
    import pwndbg.storage
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 53, in <module>
    update()
  File "/home/dc/installed/pwndbg/pwndbg/storage.py", line 50, in update
    protocol=pickle.HIGHEST_PROTOCOL)
  File "/usr/lib/python3.6/shelve.py", line 243, in open
    return DbfilenameShelf(filename, flag, protocol, writeback)
  File "/usr/lib/python3.6/shelve.py", line 227, in __init__
    Shelf.__init__(self, dbm.open(filename, flag), protocol, writeback)
  File "/usr/lib/python3.6/dbm/__init__.py", line 94, in open
    return mod.open(file, flag, mode)
_gdbm.error: [Errno 11] Resource temporarily unavailable
```

zachriggle commented:
Re: Multiple instances, fuck.  I expect we can open it read-only for secondary instances.

disconnect3d commented:
Re: Multiple instances:
Sounds like our problem: http://stackoverflow.com/questions/20420148/persistent-multiprocess-shared-cache-in-python-with-stdlib-or-minimal-dependenci

zachriggle commented:
Hmm, I'm somewhat against the idea of having a big external dependency, let me revisit this in the future.  Sorry for the long delays, I recently moved ~2500km!

MaskRay commented:
Good to know the `shelve` module.
Sadly the Berkeley DB files cannot be easily inspected.
What if some arguments have a functional dependency on others or are simply redundant?

disconnect3d commented:
@zachriggle any news on this? :<

zachriggle commented:
Nope.  Synchronized storage is hard.

One solution might be to append the current GDB `pid` to the database name.  When starting, copy the old database to the pid-appended filename.  When exiting GDB gracefully, we can just overwrite the old "original" database.

This would mean that new GDB instances always get the database of the most-recently-closed database.

It's not ideal, but a step in the right direction, and doesn't require running a "real" database (e.g. mongo / sql / etc).

zachriggle commented:
We could just use the `sqlite3` module (which handles multiple processes / connections), but then we have to deal with SQL.

zachriggle commented:
Hmm, SQLite might work.  We can just use BLOB types and have a few fields, maybe use something like peewee or PonyORM to abstract away the actual SQL stuff.

Still all-Python, since we don't need some server software running a server.

Still relatively dependency-free, since everything can be installed with `pip`

disconnect3d commented:
Speaking of ORMs there is also SqlAlchemy, but I guess we don't need such robustness. It might also be harder to learn/master, however I have no comparison as I only used it and Django ORM (and not peewee or pony).

zachriggle commented:
Apparently this is an entirely solved problem.  `joblib.memory` handles both persistent and non-persistent memoizing of Python function results concurrently across multiple processes.

https://pythonhosted.org/joblib/memory.html

disconnect3d commented:
Closing this as it is too old :(
